### DISCUSSION TITLE
Testing locally with 3rd-party cookies disabled and site data enabled

### DISCUSSION THREAD

**Discussion by vanvodk** [privacysandbox/privacy-sandbox-dev-support#118] [Phasing out third-party cookies] [answered]
I was reading this [Chrome blog post - origin trial](https://developer.chrome.com/en/blog/privacy-sandbox-unified-origin-trial/) from 2022 and try to test my current production site to see any impact. 

1. Am I correct that even though I follow all the steps, I will have to wait for new Origin Trials as previous sessions has ended? Is there a experiment flag that I can turn on locally and simulate how Chrome works like after privacy sandbox released?
2. I noticed that Chrome cookie setting "Block third-party cookies" also block site data including localStorage? Is it expected? Is it how it works after privacy sandbox released?
For example: 
There are 2 sites, **a.com** and **b.com** which is embedded in a.com as an iframe, **a.com** loaded but **b.com** did not. There is an error log saying `Failed to read the 'localStorage' property from 'Window': Access is denied for this document.` 

I know that Partitioning using "double keyed" is applied to site data but since a.com has no intentions to retrieve b.com site data. Why  did Chrome blocks site b.com to access localStorage on its own site?

**Comment by miketaylr** [comment, marked as answer]
> 2\. I noticed that Chrome cookie setting "Block third-party cookies" also block site data including localStorage? Is it expected? Is it how it works after privacy sandbox released?

Yes, this is the current expected behavior (and has been for a few years). We plan to enable partitioned storage even when 3P cookies are blocked in a future release.

**Reply by miketaylr** [reply]
FYI, as of Chrome 119, partitioned storage is enabled when 3P cookies are blocked.

**Reply by vanvodk** [reply]
Hi @miketaylr 
Will Chrome still allow users to change their cookie settings after this whole 3PC phase-out thing is rolled out (Q1 2024)? I refer to this [Allow or block third-party cookies](https://support.google.com/chrome/answer/95647#zippy=%2Callow-or-block-third-party-cookies) guideline. 

I suppose that the only change regarding [chrome://settings/cookies](url) is that the default setting will be `Block third-party cookies` instead of `Block third-party cookies in Incognito mode`. Is it a correct statement?

**Reply by miketaylr** [reply]
Hi @vanvodk - we should have more to announce on this soon. Thanks.

**Comment by krgovind** [comment]
> 1. Am I correct that even though I follow all the steps, I will have to wait for new Origin Trials as previous sessions has ended? Is there a experiment flag that I can turn on locally and simulate how Chrome works like after privacy sandbox released?

Yes, every API does indeed have a flag that you can enable to test locally. I would recommend looking at the developer article corresponding to the API. For example, see [First-Party Sets testing instructions](https://developer.chrome.com/en/blog/first-party-sets-testing-instructions/#how-to-test-locally).

> 2. I noticed that Chrome cookie setting "Block third-party cookies" also block site data including localStorage? Is it expected? Is it how it works after privacy sandbox released?

@miketaylr shared our expected shipping behavior in a previous comment. I wanted to add that in order to enable developers to test third-party cookie blocking with storage partitioning enabled, we are in the process of developing a command-line switch for local testing. Please see the tracking bug: [crbug.com/1441039](https://crbug.com/1441039). The flag is available starting M115, but we are in the process of adding the storage partitioning behavior to it. We're hoping to make this available starting Chrome 116. The change should be available on the dev channel in a couple of weeks.

**Reply by vanvodk** [reply]
@krgovind Thanks for the article! Can you please also explain a little more about my concerns as follows?

1. Is it true that the rationale, which "Block third-party cookies" also block site data, because localStorage is accessible between frames? Therfore even **top** frame or  embedded **iframe** will not able to read/write data into site data?

> @miketaylr shared our expected shipping behavior in a previous comment. I wanted to add that in order to enable developers to test third-party cookie blocking with storage partitioning enabled, we are in the process of developing a command-line switch for local testing. Please see the tracking bug: [crbug.com/1441039](https://crbug.com/1441039). The flag is available starting M115, but we are in the process of adding the storage partitioning behavior to it. We're hoping to make this available starting Chrome 116. The change should be available on the dev channel in a couple of weeks.

2. With Storage partitioning turned on, site data will be secured with double-keyed domains, therefore it will then allow **top** frame or  embedded **iframe** to access to site data?

I followed this [test guideline](https://developer.chrome.com/en/blog/privacy-sandbox-unified-origin-trial/) but noticed nothing has changed nor applied. I'm going to try the testing instructions article you just sent

**Reply by krgovind** [reply]
1. The current "Block third-party cookies" does block site data (or "storage") in cross-site iframes (e.g. `b.com` embedded on `a.com`), but we are planning to change that in a upcoming release so that `b.com` has access to [partitioned storage](https://developer.chrome.com/en/docs/privacy-sandbox/storage-partitioning/). Does that answer your question?
2. Could you clarify which API you are trying to test? The article is shared is specifically for First-Party Sets, but if you're looking to test one of the ads measurement/relevance APIs, [this section of the article you linked to](https://developer.chrome.com/en/blog/privacy-sandbox-unified-origin-trial/#local-testing) is indeed the correct one to follow.

**Reply by vanvodk** [reply]
Yes indeed. Thanks for your help!

**Reply by vanvodk** [reply]
@krgovind I just happened to see that the [crbug.com/1441039](https://crbug.com/1441039) is Fixed (Close). Is there any release channel that I can use to test it out? I used Chrome Canary to test it out but it doesn't seem to work.

**Reply by krgovind** [reply]
The flag is available starting Chrome 115. It did get recently renamed to `--test-third-party-cookie-phaseout` though. How are you using it on the command-line exactly?

**Reply by vanvodk** [reply]
I ran this command on Windows Terminal

```
--test-third-party-cookie-phaseout --enable-features="FirstPartySets,StorageAccessAPI,StorageAccessAPIForOriginExtension,PageInfoCookiesSubpage,PrivacySandboxFirstPartySetsUI" `
--use-first-party-set="{\"primary\": \"https://first-party-sets.glitch.me\", \"associatedSites\": [\"https://fps-member-1.glitch.me\"]}" `
https://first-party-sets.glitch.me/
```

and the Google test sites shown me this

![image](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support/assets/11583191/68ec3ee5-c845-4e5d-91a8-e1a29774cf47)

I also got this error log from the developer console after clicking on "Click for cookies" option. I suppose that `requestStorageAccessFor` should not return any exception. Is it a correct statement? Or am I misunderstanding?

![image](https://github.com/GoogleChromeLabs/privacy-sandbox-dev-support/assets/11583191/b0393d0b-ecdd-4b97-abbc-315bbef76c7e)

**Reply by DCtheTall** [reply]
Hello @vanvo1412 thank you for filing this issue and apologies you are experiencing issues.

For the first screenshot, when I run Chrome Canary on version 117 with the command line you provided, I see "Third-pary cookies look like they're blocked" display a [PASS]. Could you please verify your Chrome version?

The second screenshot is indeed a bug in the command line switch functionality, thank you for pointing it out. I am working on a [patch](https://chromium-review.googlesource.com/c/chromium/src/+/4678172) and until then, you can enable FPS by opening chrome://settings/cookies and after selecting "Block Third-Party Cookies" and clicking the "Allow related sites to see activity in the group" toggle will enable FPS.

**Comment by vanvodk** [comment]
Hello @DCtheTall, indeed I run the same Chrome Canary on version 117. But I was able to make it work using Chrome Beta on version 115. I expect Chrome Canary will get updated more frequently than Chrome Beta, and it will get all the patches that you mentioned there. 

Even I do this, it does not work for until I tried Chrome Beta instead of Chrome Canary. In summary, it all works out when I use Chrome Beta
> you can enable FPS by opening chrome://settings/cookies and after selecting "Block Third-Party Cookies" and clicking the "Allow related sites to see activity in the group" toggle will enable FPS.

**Reply by DCtheTall** [reply]
Hey @vanvodk thanks for verifying your version. It is indeed strange to us that it is working in Beta and not Canary. After thinking it over, here are a couple more things which may help us figure out the cause:

Can you send a screenshot of your chrome://settings/cookies pages for both your Canary and Beta instances?

Can you ensure your Canary does not have any settings allowlisting any domains from sending cookies when 3PC blocking is on? These allowlists in settings take priority over the 3PC blocking setting (see "Allow or block cookies for a specific site" [on this page](https://support.google.com/chrome/answer/95647)).

**Reply by vanvodk** [reply]
Hello @DCtheTall, my bad. Actually Chrome Beta and Canary work the same as expected. `"Block Third-Party Cookies"` and  `"Allow related sites to see activity in the group"` were enabled in Chrome Canary while they are not in Chrome Beta. Here is the correction

1. `"Block Third-Party Cookies"` setting blocks both 3PC and browser storage while `--test-third-party-cookie-phaseout` only blocks 3PC
2. Both `"Block Third-Party Cookies"` and `--test-third-party-cookie-phaseout` throw the same exception on the test site. I expect I would see cross-site cookies after I press the button "Click for cookies", is it a correct statement?
> ![image](https://user-images.githubusercontent.com/11583191/252472615-b0393d0b-ecdd-4b97-abbc-315bbef76c7e.png)
